### PR TITLE
Workaround topoaa

### PR DESCRIPTION
**Pull request by VGPReys**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and that you comply with the following criteria:

- [ ] You have sticked to Python. Please talk to us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [ ] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [ ] You structured the code into small functions as much as possible. You can use classes if there is a (state) purpose
- [ ] Your code follows our coding style
- [ ] You wrote tests for the new code
- [ ] `tox` tests pass. *Run `tox` command inside the repository folder*
- [ ] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [ ] PR does not add any dependencies, unless permission granted by the HADDOCK team
- [ ] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

This PR try to find a path to be able to do a haddock3 run without having to start with `topoaa`.

Basically, now the input molecules are handled by the `ModuleIO` class (in `haddock/libs/libontology.py`), that mimic the output of an `io.json`.
Small modifications had to be applied to the `topoaa` module to fit this new behavior.

### CODE REVIEW

File: src/haddock/clis/cli_score.py [hunk 1]
```diff
@@ -103,7 +103,64 @@ def cli(ap: ArgumentParser, main: Callable[..., None]) -> None:
...
+                )
+        if value != default_emscoring[param]:
+            print(
+                f"* ATTENTION * Value ({value}) of parameter {param} "
+                f"different from default ({default_emscoring[param]})"
+                )
+            # get the type of default value
+            default_type = type(default_emscoring[param])
+            # convert the value to the same type
+            if default_type == bool:
```
Comment (rvhonorato) [inline src/haddock/clis/cli_score.py:145]: the way to check for a type is `isinstance(var, type)`

File: src/haddock/core/typing.py [hunk 2]
```diff
@@ -30,6 +30,7 @@
     Generic,
     Iterable,
     Iterator,
+    List,
```
Comment (rvhonorato) [inline src/haddock/core/typing.py:33]: `List` is not used

File: src/haddock/gear/prepare_run.py [hunk 3]
```diff
@@ -315,6 +315,12 @@ def setup_run(
             dec_all=True,
         )
 
+    first_module_id = list(modules_params.keys())[0]
+    if (topoaa_module_id := "topoaa.1") in modules_params.keys():
+        topology_params = modules_params[topoaa_module_id]
+    else:
+        topology_params = {}
```
Comment (rvhonorato) [inline src/haddock/gear/prepare_run.py:322]: I see in the old version this `topoaa.1` was also present, can you explain why?

File: src/haddock/gear/prepare_run.py [hunk 4]
```diff
@@ -327,24 +333,27 @@ def setup_run(
     else:
         copy_molecules_to_topology(
             general_params["molecules"],
-            modules_params["topoaa.1"],
+            topology_params,
         )
+        # copy_molecules_to_topology(
+        #     general_params["molecules"],
+        #     modules_params[first_module_id],
+        # )
```
Comment (rvhonorato) [inline src/haddock/gear/prepare_run.py:342]: Remove the comented code

File: src/haddock/gear/prepare_run.py [hunk 5]
```diff
@@ -807,40 +821,48 @@ def copy_molecules_to_data_dir(
         Whether to preprocess input molecules. Defaults to ``True``.
         See :py:mod:`haddock.gear.preprocessing`.
     """
-    topoaa_dir = zero_fill.fill("topoaa", 0)
+    # Removes digit from module name
+    # Build regex to capture '<name>.<digit>'
+    name_digit_regex = re.compile(r"(\w+)\.\d+")
```
Comment (rvhonorato) [inline src/haddock/gear/prepare_run.py:826]: This will not capture with the module name does not have a digit, is that the intended behaviour? If you want to capture names that might not have the `.`, the regex should be `(\w+)(\.\d+)?`

File: src/haddock/gear/prepare_run.py [hunk 6]
```diff
@@ -789,8 +800,11 @@ def copy_molecules_to_topology(
 
 
 def copy_molecules_to_data_dir(
-    data_dir: Path, topoaa_params: ParamMap, preprocess: bool = True
-) -> None:
+        data_dir: Path,
+        topoaa_params: ParamMap,
+        _first_module_name: str,
```
Comment (rvhonorato) [inline src/haddock/gear/prepare_run.py:805]: missing `_first_module_name` in the docstring

File: src/haddock/gear/prepare_run.py [hunk 7]
```diff
@@ -1053,7 +1075,10 @@ def populate_topology_molecule_params(topoaa: ParamMap) -> None:
     return
 
 
-def populate_mol_parameters(modules_params: ParamMap) -> None:
+def populate_mol_parameters(
+        modules_params: ParamMap,
+        topology_params: ParamMap,
```
Comment (rvhonorato) [inline src/haddock/gear/prepare_run.py:1080]: missing `topology_params` in the docstring

File: src/haddock/gear/prepare_run.py [hunk 8]
```diff
@@ -1079,7 +1104,7 @@ def populate_mol_parameters(modules_params: ParamMap) -> None:
         Alter the dictionary in place.
     """
     # the starting number of the `mol_` parameters is 1 by CNS definition.
-    num_mols = range(1, len(modules_params["topoaa.1"]["molecules"]) + 1)
+    num_mols = range(1, len(topology_params["molecules"]) + 1)
```
Comment (rvhonorato) [inline src/haddock/gear/prepare_run.py:1107]: Why can't you just re-use `modules_params` here?

File: src/haddock/libs/libontology.py [hunk 9]
```diff
@@ -233,11 +255,226 @@ def remove_missing(self) -> None:
...
+            f"No molecules could be found in `{input_molecules_dir}`"
+        # Sort them by creation date (which is also input order)
+        input_molecules.sort(key=getmtime)  # FIXME: getctime ?
+        # Set input attribute
+        self.input = input_molecules
+
+        # Set parsing variables
+        molecules_list: list[dict[int, PDBFile]] = [
+            Molecule(input_file).pdb_files
+            for input_file in self.input
```
Comment (rvhonorato) [inline src/haddock/libs/libontology.py:289]: `input_file` could be either a `dict` or a `PDBFile` - this will fail if it's not a dict

File: src/haddock/libs/libontology.py [hunk 10]
```diff
@@ -233,11 +255,226 @@ def remove_missing(self) -> None:
...
+            ]
+        # And fake them to be the output of the previous io
+        self.output = molecules_list
+    
+
+class Molecule:
```
Comment (rvhonorato) [inline src/haddock/libs/libontology.py:295]: Why did you move this class here?

File: src/haddock/libs/libpdb.py [hunk 11]
```diff
@@ -110,24 +113,64 @@ def get_supported_residues(haddock_topology: FilePath) -> list[str]:
...
+    models_starts: int = 0
+    model_ends: int = 0
+    with open(pdb_file_path, 'r') as filin:
+        for line in filin:
+            if line.startswith("ENDMDL"):
+                model_ends += 1
+            elif line.startswith("MODEL"):
+                models_starts += 1
+    nb_models = max(1, model_ends)
+    assert max(1, models_starts) == nb_models
```
Comment (rvhonorato) [inline src/haddock/libs/libpdb.py:172]: Why this assert here?

File: src/haddock/libs/libontology.py [hunk 12]
```diff
@@ -233,11 +255,226 @@ def remove_missing(self) -> None:
...
+        """Load first molecules at the stat of the workflow.
+
+        Parameters
+        ----------
+        input_molecules_dir : Path
+            Directory where the input molecules are stored.
+        """
+        # Gather all input molecules
+        input_molecules = list(input_molecules_dir.glob('*.pdb'))
+        assert input_molecules != [], \
```
Comment (rvhonorato) [inline src/haddock/libs/libontology.py:279]: I think here would be best to raise an error instead of the assert

File: src/haddock/modules/topology/topoaa/__init__.py [hunk 13]
```diff
@@ -173,79 +117,74 @@ def _run(self) -> None:
...
                     self.params,
                     parameters_for_this_molecule,
                     default_params_path=self.toppar_path,
                     write_to_disk=not self.params["less_io"],
-                )
-
-                self.log("Topology CNS input created")
+                    )
+                self.log(
+                    f"Topology CNS input created in {topoaa_input.name}"
```
Comment (rvhonorato) [inline src/haddock/modules/topology/topoaa/__init__.py:158]: `generate_topology` might return a `str`, so this `.name` will fail

File: tests/test_gear_yaml2cfg.py [hunk 14]
```diff
@@ -97,3 +112,19 @@ def test_yaml2cfg_test_no_header():
         shallow=False,
         )
     p.unlink()
+
+
+def test_yaml_duplicated_params(default_yaml_files):
```
Comment (rvhonorato) [inline tests/test_gear_yaml2cfg.py:117]: Isn't this being done in https://github.com/haddocking/haddock3/pull/913?

File: tests/test_libontology.py [hunk 15]
```diff
@@ -81,6 +86,74 @@ def io_json_file(io_data) -> Generator[Path, None, None]:
         yield Path(f.name)
 
 
+@pytest.fixture
+def molecule():
+    return Molecule(None)
```
Comment (rvhonorato) [inline tests/test_libontology.py:91]: Strange this is not raising a type error, but I see your point, just add a  `# type: ignore`

File: tests/test_libworkflow.py [hunk 16]
```diff
@@ -19,13 +20,14 @@ def test_WorkflowManager(caplog):
...
-            ParamDict,
-            start=0,
-            other_params=Any,
-            )
-        workflow.postprocess()
-        first_log_line = str(caplog.records[0].message)
-        second_log_line = str(caplog.records[1].message)
-        assert first_log_line == "Reading instructions step 0_topoaa"
-        assert second_log_line == "Running haddock3-analyse on ./, modules [], with top_cluster = 10"  # noqa : E501
+        with working_directory(_tmpdir):
```
Comment (rvhonorato) [inline tests/test_libworkflow.py:23]: Why do you need to change the context here?

File: integration_tests/test_alascan.py [hunk 17]
```diff
@@ -17,7 +17,7 @@ def alascan_module():
     """Return a default alascan module."""
     with tempfile.TemporaryDirectory(dir=".") as tmpdir:
         alascan = AlascanModule(
-            order=0, path=".", initial_params=DEFAULT_ALASCAN_CONFIG
+            order=1, path=Path("."), initial_params=DEFAULT_ALASCAN_CONFIG
```
Comment (rvhonorato) [inline integration_tests/test_alascan.py:20]: Why do you need to change the order here? Also in the other tests

File: integration_tests/test_topoaa.py [hunk 18]
```diff
@@ -1,31 +1,58 @@
...
+        initial_params=DEFAULT_TOPOAA_CONFIG,
         )
-        topoaa.__init__(path=tmpdir, order=0)
-        topoaa.params["molecules"] = [
-            Path(DATA_DIR, "docking-protein-protein/data/e2aP_1F3G.pdb"),
-            Path(DATA_DIR, "docking-protein-protein/data/hpr_ensemble.pdb"),
-        ]
-        topoaa.params["mol1"] = {"prot_segid": "A"}
-        topoaa.params["mol2"] = {"prot_segid": "B"}
+    #topoaa.__init__(path=modulename_path, order=0)
```
Comment (rvhonorato) [inline integration_tests/test_topoaa.py:48]: remove comented code

File: src/haddock/clis/cli_score.py [hunk 19]
```diff
@@ -103,7 +103,64 @@ def cli(ap: ArgumentParser, main: Callable[..., None]) -> None:
 
 def maincli() -> None:
     """Execute main client."""
-    cli(ap, main)
+    cli(_ap(), main)
+
+
+def get_parameters(kwargs: dict[str, Any]) -> dict[str, Any]:
```
Comment (rvhonorato) [inline src/haddock/clis/cli_score.py:109]: No test for this function

**Review by rvhonorato — CHANGES_REQUESTED** [review]
It's not simple to untangle this part of the code, as you probably noticed - great job getting into it

File: src/haddock/gear/prepare_run.py (hunk 3 above)
Comment (VGPReys) [reply src/haddock/gear/prepare_run.py:322]: Required by the function `populate_topology_molecule_params()`, to map the `molX` parameters info to input molecules, in the case `topoaa` is the first module.
Without this, we loose the information.

Comment (VGPReys) [reply src/haddock/gear/prepare_run.py:322]: Added comments on the code directly

File: src/haddock/gear/prepare_run.py (hunk 5 above)
Comment (VGPReys) [reply src/haddock/gear/prepare_run.py:826]: All of them have a '.' at this stage, separating `module name` and `module step`.
This is performed/ensured earlier by `load()` from `haddock.gear.config`.
But indeed, your regex is more flexible.

File: src/haddock/gear/prepare_run.py (hunk 8 above)
Comment (VGPReys) [reply src/haddock/gear/prepare_run.py:1107]: Because now molecules are stored in the `topology_params["molecules"]` instead of modules params.
We cannot point to the first module parameters, as they do not have the `molecules` parameter defined in their `defaults.yaml` and therefore not valid later...
Not easy...

Comment (VGPReys) [reply src/haddock/gear/prepare_run.py:1107]: In the case of `topoaa`, we have defined the parameter `molecules`, with expert level `hidden`...
An other trick

File: src/haddock/libs/libontology.py (hunk 10 above)
Comment (VGPReys) [reply src/haddock/libs/libontology.py:295]: Because of recursive import issues...
`libontology` is:
- defining `PDBFile` class
- Requiring the use of the `Molecule` class

Before, the `Molecule` class was defined in `structure.py`, but now, is asking for the type `PDBFile` that is defined in `libontology`.

So, to my knowledge, moving it here was the only solution.

File: src/haddock/libs/libpdb.py (hunk 11 above)
Comment (VGPReys) [reply src/haddock/libs/libpdb.py:172]: In the case of an ensemble: I am counting `ENDMDL` statements to define the number of models, and I just want to make sure that the number of `ENDMDL` statements is equal to the number of `MODEL` statements.
Otherwise, there should not be any `ENDMDL` nor `MODEL` statements, and I return `1`

File: tests/test_gear_yaml2cfg.py (hunk 14 above)
Comment (VGPReys) [reply tests/test_gear_yaml2cfg.py:117]: I guess I messed up between PRs...

File: tests/test_libworkflow.py (hunk 16 above)
Comment (VGPReys) [reply tests/test_libworkflow.py:23]: otherwise the temporary directory is simply not used.

File: integration_tests/test_alascan.py (hunk 17 above)
Comment (VGPReys) [reply integration_tests/test_alascan.py:20]: This is an issue related to this new implementation, and the inheritance of `AlascanModule` <- `HaddockModule.__init__()`
- If the order is `0` (first module in the workflow): It will search for input molecules stored somewhere that is not created for this integration test and `[src/haddock/libs/libontology.py](https://github.com/haddocking/haddock3/pull/921/files/38bb99b82ddf685e8f56d2da13e007821cbf6638#diff-15b98e8f9e924d3f27cf4aa6d3afebfcea090bc0a6594271706e7e134e3db4b2) line 279 will trigger a crash.
- If not, it will not search for it, and all fine

File: src/haddock/libs/libontology.py (hunk 9 above)
Comment (VGPReys) [reply src/haddock/libs/libontology.py:292]: In this case, `input_file` is a `Path` to the input molecules from the config file.

File: src/haddock/core/typing.py (hunk 2 above)
Comment (VGPReys) [reply src/haddock/core/typing.py]: Used in `libontology`.
Imported by other scripts.

**Comment by mgiulini** [comment]
I tested the PR: it works well when the workflow is made of non-CNS modules, but when CNS modules are included a workflow without topoaa badly fails at the CNS preparation steps, without catching the error. Here an example output of [contmap-test](https://github.com/haddocking/haddock3/blob/main/examples/analysis/contmap-test.cfg) removing topoaa:
```
(haddock3) UU-CW4VKWDG2H:analysis Giuli003$ haddock3 contmap-test.cfg 
[2024-07-26 15:48:10,564 cli INFO] 
##############################################
#                                            #
#                 HADDOCK 3                  #
#                                            #
##############################################

Starting HADDOCK 3.0.0 on 2024-07-26 15:48:00

Python 3.9.18 (main, Sep 11 2023, 08:25:10) 
[Clang 14.0.6 ]

[2024-07-26 15:48:14,905 libworkflow INFO] Reading instructions step 0_rigidbody
[2024-07-26 15:48:14,905 libworkflow INFO] Reading instructions step 1_clustfcc
[2024-07-26 15:48:14,905 libworkflow INFO] Reading instructions step 2_contactmap
[2024-07-26 15:48:15,343 base_cns_module INFO] Running [rigidbody] module
[2024-07-26 15:48:15,344 __init__ INFO] [rigidbody] crossdock=true
[2024-07-26 15:48:15,344 __init__ INFO] [rigidbody] Preparing jobs...
[2024-07-26 15:48:15,344 libutil INFO] Selected 5 cores to process 20 jobs, with 8 maximum available cores.
[2024-07-26 15:48:15,354 libparallel INFO] Using 5 cores
Process Worker-1:
Process Worker-2:
Process Worker-3:
Process Worker-4:
Traceback (most recent call last):
Traceback (most recent call last):
  File "/Users/Giuli003/anaconda3/envs/haddock3/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/Users/Giuli003/software/haddock3/src/haddock/libs/libparallel.py", line 88, in run
    r = task.run()
  File "/Users/Giuli003/software/haddock3/src/haddock/libs/libparallel.py", line 72, in run
    return self.function(*self.args, **self.kwargs)
Traceback (most recent call last):
  File "/Users/Giuli003/software/haddock3/src/haddock/libs/libcns.py", line 307, in prepare_cns_input
    raise ValueError(f"Topology not found for pdb {pdb.rel_path}.")
  File "/Users/Giuli003/anaconda3/envs/haddock3/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
ValueError: Topology not found for pdb ../data/0_rigidbody/1a2k_r_u.pdb.
```

**Comment by rvhonorato** [comment]
The error above means you are trying to execute a CNS module without having generated the topologies, not related to the contact module

Removing `topoaa` from being the "module 0" does add this dependency

**Comment by amjjbonvin** [comment]
Basically rigidbody in this case can not be run if topoaa was not first run.

Need to define dependencies...

**Comment by mgiulini** [comment]
yes, precisely. we need to catch this exception at the beginning asking the user to add topoaa to the workflow

**Comment by VGPReys** [comment]
Thanks for the review.
I will make some modifications to solve this problem.

**Comment by rvhonorato** [comment]
Defining this dependency graph is quite complex and definitely besides the scope of this pr, could you please handle this in another?

Remember this is a beta version and these kinds of uncaught exceptions are tolerable, it's a work in progress anyway :)

**Review by mgiulini — APPROVED** [review]
